elastic/kibana
# Don't trigger auto-refresh until previous refresh completes

## Summary

Closes https://github.com/elastic/kibana/issues/86947
Partially addresses https://github.com/elastic/kibana/issues/76521
Unblocks https://github.com/elastic/kibana/issues/95643


This pr changes how auto-refresh works: now when auto-refresh observable emits app needs to call a `done()` callback notifying the auto-refresh services that it is OK to start a new refresh loop. 

Each app decides slightly differently what "complete" means: 

1. Dashboard and Lens use underlying search session services. Assume it is OK to start a next auto-refresh loop when the next session completes
2. Discover awaits $scope.fetch() 
3. Visualize awaits expressionHandler.update()

There are likely edge cases with these "complete" estimations, but in the end, it isn't really important if `done()` is called too often or too soon.  


### How to test

Makes sense to only test Discover, Visualize, Lens and Dashboard because only they use `getAutoRefreshFetch$`. 
You can:

1. Slower your network using custom present with 5s latency 
2. Set a timed refresh in 3s interval
3. You should see something like this in network table:
```
0s ---- requests
....
5s ---- request finishes 
....
8s ---- request starts
```


Please note,
I noticed that map embeddable doesn't use `getAutoRefreshFetch$`, but is using a custom `window.interval` so it behaves oddly on a dashboard with a "slowed down" network. (this is not a regression, but a bug). I will open a separate issue for the geo team.

### Checklist


### For maintainers
